pkg/testutils/fingerprintutils/fingerprint.go

```diff
@@ -0,0 +1,230 @@
+
+	for rows.Next() {
+		var name string
+		var id uint32
+		if err := rows.Scan(&id, &name); err != nil {
+			return nil, fmt.Errorf("error scanning table_name for db %s: %w", dbName, err)
+		}
+		tableIDs = append(tableIDs, id)
+		idToName[id] = name
+	}
+	if err := rows.Err(); err != nil {
+		return nil, fmt.Errorf("error iterating over table_name rows for database %s: %w", dbName, err)
+	}
+	nameToFingerprint := make(map[string]int64, len(tableIDs))
+	fingerprints, err := FingerprintTables(ctx, db, tableIDs, optFuncs...)
+	if err != nil {
+		return nil, fmt.Errorf("failed fingerprint tables in database %s:%w", dbName, err)
+	}
+	for i, fingerprint := range fingerprints {
+		id := tableIDs[i]
+		name := idToName[id]
+		nameToFingerprint[name] = fingerprint
+	}
+	return nameToFingerprint, err
+}
+
+// FingerprintCluster fingerprints
+func FingerprintCluster(
+	ctx context.Context, db *gosql.DB, includeSystemDB bool, optFuncs ...func(*FingerprintOption),
+) (map[string]map[string]int64, error) {
+
+	dbNames := make([]string, 0)
+	rows, err := db.QueryContext(ctx, `SELECT database_name FROM [SHOW DATABASES]`)
+	if err != nil {
+		return nil, errors.New("could not get database names")
+	}
+	defer rows.Close()
+	for rows.Next() {
+		var name string
+		if err := rows.Scan(&name); err != nil {
+			return nil, fmt.Errorf("error scanning for db names: %w", err)
+		}
+		dbNames = append(dbNames, name)
+	}
+	if err := rows.Err(); err != nil {
+		return nil, fmt.Errorf("error iterating over database names: %w", err)
+	}
+	dbFingerprints := make(map[string]map[string]int64)
+	for _, dbName := range dbNames {
+		if dbName == "system" && !includeSystemDB {
+			continue
+		}
+		dbFingerprint, err := FingerprintDatabase(ctx, db, dbName, optFuncs...)
+		if err != nil {
+			return nil, err
+		}
+		dbFingerprints[dbName] = dbFingerprint
+	}
+	return dbFingerprints, nil
+}
+
+func CompareDatabaseFingerprints(db1, db2 map[string]int64) error {
+	for tableName, tableFingerprint := range db1 {
+		table2Fingerprint, ok := db2[tableName]
+		if !ok {
+			return errors.Newf("%s table not in second database", tableName)
+		}
+		if tableFingerprint != table2Fingerprint {
+			return errors.Newf("fingerprint mismatch on %s table: %d != %d", tableName,
+				tableFingerprint, table2Fingerprint)
+		}
+		delete(db2, tableName)
+	}
+	if len(db2) > 0 {
+		return errors.Newf("second database has more tables %s", db2)
+	}
+	return nil
+}
+
+func CompareClusterFingerprints(c1, c2 map[string]map[string]int64) error {
+	for dbName, dbFingerprints := range c1 {
+		db2Fingerprints, ok := c2[dbName]
+		if !ok {
+			return errors.Newf("%s table not in second database", dbName)
+		}
+		if err := CompareDatabaseFingerprints(dbFingerprints, db2Fingerprints); err != nil {
+			return fmt.Errorf("failed comparing fingerprints for database %s: %w", dbName, err)
+		}
+		delete(c2, dbName)
+	}
+	if len(c2) > 0 {
+		return errors.Newf("second cluster has more databases %s", c2)
+	}
+	return nil
+}
```

Comment:
Fine to keep but I wonder how much clearer this output is than `require.Equal` on the two input maps.

Reply:
Sure, I more meant whether we need this custom comparison function at all.

Reply:
Ah i see. I like this func because it will point you to exactly what's wrong with your fingerprint comparison. require over the maps implies that the test debugger has to squint at printed maps :D.

I could also see a future where we scan over the mismatched table to find the wrong key.